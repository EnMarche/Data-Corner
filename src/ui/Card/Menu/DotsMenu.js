import React, { useState } from 'react'
import { IconButton, Menu, MenuItem as MuiMenuItem } from '@mui/material'
import MoreVertIcon from '@mui/icons-material/MoreVert'
import { styled } from '@mui/system'
import PropTypes from 'prop-types'
import Loader from 'ui/Loader'

const Wrapper = styled('div')`
  display: flex;
`

const MenuItem = styled(MuiMenuItem)`
  font-size: 13px;
  padding: ${({ theme }) => theme.spacing(0.5, 1)};
  border-width: 1px;
  border-style: solid;
  border-color: ${({ theme }) => theme.palette.gray100};
  border-radius: 8.35px;
  background-color: ${({ theme }) => theme.palette.whiteCorner};
  &:hover {
    background-color: ${({ theme }) => theme.palette.gray100};
  },
`

export const DotsMenuItem = ({ onClick, closeMenu, loader = false, children }) => {
  const handleClick = async () => {
    await onClick()
    closeMenu()
  }
  return (
    <MenuItem onClick={handleClick}>
      {loader && (
        <>
          <Loader size={12} />
          &nbsp;
        </>
      )}
      {children}
    </MenuItem>
  )
}

DotsMenuItem.propTypes = {
  onClick: PropTypes.func.isRequired,
  closeMenu: PropTypes.func,
  loader: PropTypes.bool,
  children: PropTypes.node.isRequired,
}

const DotsMenu = ({ children }) => {
  const [anchorEl, setAnchorEl] = useState(null)

  const handleClose = () => {
    setAnchorEl(null)
  }

  const handleButtonClick = e => {
    setAnchorEl(e.currentTarget)
  }

  const closeMenu = () => {
    setAnchorEl(null)
  }

  return (
    <Wrapper>
      <IconButton size="small" onClick={handleButtonClick} sx={{ p: 0 }}>
        <MoreVertIcon />
      </IconButton>
      <Menu
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={handleClose}
        sx={{
          fontSize: '13px',
          padding: 0,
          borderRadius: '8.35px',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
      >
        {React.Children.map(children, c => React.cloneElement(c, { closeMenu }))}
      </Menu>
    </Wrapper>
  )
}

DotsMenu.propTypes = {
  children: PropTypes.node.isRequired,
}

export default DotsMenu
