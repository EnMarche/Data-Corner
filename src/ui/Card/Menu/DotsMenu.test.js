import { fireEvent, render, screen } from '@testing-library/react'
import DotsMenu, { DotsMenuItem } from 'ui/Card/Menu/DotsMenu'

jest.mock('@mui/system', () => ({
  styled: c => () => c,
}))
jest.mock('@mui/icons-material/MoreVert', () => () => <div className="mui-MoreVertIcon-mock" />)
jest.mock('@mui/material', () => ({
  Button: ({ children }) => <div className="mui-button-mock">{children}</div>,
  IconButton: ({ children }) => <div className="mui-button-mock">{children}</div>,
  Menu: ({ children }) => <div className="mui-Menu-mock">{children}</div>,
  MenuItem: ({ children, onClick }) => (
    <div className="mui-MenuItem-mock" onClick={onClick} data-testid="mui-MenuItem-mock">
      {children}
    </div>
  ),
}))
describe('DotsMenu', () => {
  it('displays a DotsMenu', () => {
    const mockOnClick = jest.fn()
    const { container } = render(
      <DotsMenu>
        <DotsMenuItem onClick={mockOnClick}>foo</DotsMenuItem>
      </DotsMenu>
    )
    expect(container).toMatchSnapshot()
  })
  it('calls a mock when clicking on menu item', () => {
    const mockOnClick = jest.fn()
    render(
      <DotsMenu>
        <DotsMenuItem onClick={mockOnClick}>foo</DotsMenuItem>
      </DotsMenu>
    )

    const menuItem = screen.getByTestId('mui-MenuItem-mock')
    fireEvent.click(menuItem)
    expect(mockOnClick).toHaveBeenCalled()
  })
})
