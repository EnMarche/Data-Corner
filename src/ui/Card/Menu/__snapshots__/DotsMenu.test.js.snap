// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DotsMenu displays a DotsMenu 1`] = `
<div>
  <div>
    <div
      class="mui-button-mock"
    >
      <div
        class="mui-MoreVertIcon-mock"
      />
    </div>
    <div
      class="mui-Menu-mock"
    >
      <div
        class="mui-MenuItem-mock"
        data-testid="mui-MenuItem-mock"
      >
        foo
      </div>
    </div>
  </div>
</div>
`;
